server/handler/query.go

```diff
@@ -74,8 +74,17 @@ func Query(db service.SQLDB) RequestProcessFunc {
 		c := make(chan error, 1)
 
 		var rows *sql.Rows
+		conn, err := db.Conn(r.Context())
+		defer conn.Close()
```

Comment:
If we close after each query, does it negate the `GITBASEPG_CONN_MAX_LIFETIME` config? I mean that the connection pool will not contain idle connections ever, right?

Reply:
> If we close after each query, does it negate the `GITBASEPG_CONN_MAX_LIFETIME` config?

It shouldn't. By reading the docs and the article I linked [here](https://github.com/src-d/gitbase-web/pull/306#discussion_r250685778), the connection must be closed in order to make it available again in the pool. From the doc:

> Every Conn must be returned to the database pool after use by calling Conn.Close.

I *guess* that without closing it, it just becomes some sort of opened zombie connection. This leak could lead the pool to consider all the connections as "in-use" and reach the limit number of opened connections. By considering them "in-use" I also think that the connection won't be considered expired and won't honour `GITBASEPG_CONN_MAX_LIFETIME`.

Anyway I'm not sure about all this, but maybe I can try doing a simple test where I open a connection without closing it and see the [stats of the pool](https://golang.org/pkg/database/sql/#DBStats).

Reply:
Thanks, I was assuming a different meaning for `Conn.Close()` :+1:

Reply:
Okay, just to be sure I tested id 😛. Here's the code that I ran:

```package main

import (
	"database/sql"
	"fmt"
	"net/http"

	_ "github.com/mattn/go-sqlite3"
)

func main() {
	db, err := sql.Open("sqlite3", "./foo.db")
	db.SetMaxOpenConns(3)

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		fmt.Println("===START===")
		fmt.Fprintf(w, "Welcome to my website!")
		if err != nil {
			fmt.Println(err)
			return
		}

		fmt.Printf("Stats: %v\n", db.Stats())
		fmt.Println("Opening connection...")

		conn, err := db.Conn(r.Context())
		if err != nil {
			fmt.Println(err)
			return
		}

		fmt.Printf("Stats: %v\n", db.Stats())
		fmt.Println("Getting connection from pool")

		fmt.Println("Connection opened")
		fmt.Println("Pinging connection...")
		conn.PingContext(r.Context())
		fmt.Println("Avoided closing connection")
		fmt.Println("===END===")
	})

	http.ListenAndServe(":80", nil)
}
```

And here's the screen of the ouput:

<img width="518" alt="screenshot 2019-01-28 at 19 15 08" src="https://user-images.githubusercontent.com/5599208/51856907-3325dd80-2331-11e9-9730-0a219a31517e.png">

After the 3rd request, the server is not able to open new connections, and the request hangs indefinitely. I tried to add the `defer conn.Close()` and the stats shows the swtich between idle and in use at each request.

Reply:
Actually in our case I don't think we have a limit on the number of connections, but it's clearly a leak 👍

Reply:
Thank you for the thorough follow-up. In our case we should have the [default number of idle connections, 2](https://golang.org/pkg/database/sql/#DB.SetMaxIdleConns).